### PR TITLE
RavenDB-13545 provide custom S3 hostname for backups

### PR DESCRIPTION
**Pull request by haludi**
https://issues.hibernatingrhinos.com/issue/RavenDB-13545

### CODE REVIEW

File: src/Raven.Server/Documents/PeriodicBackup/Aws/RavenAwsS3Client.cs [hunk 1]
```diff
@@ -43,9 +47,19 @@ public RavenAwsS3Client(S3Settings s3Settings, Progress progress = null, Logger
...
+                if (string.IsNullOrWhiteSpace(s3Settings.AwsRegionName))
+                    throw new ArgumentException("AWS region Name cannot be null or empty");
+                AwsRegion = s3Settings.AwsRegionName.ToLower();
+            }
+            else
+            {
+                if (string.IsNullOrWhiteSpace(s3Settings.AwsRegionName) == false)
+                    throw new ArgumentException($"There is no need for region in custom S3 server: AwsRegionName({s3Settings.AwsRegionName})");
+                _customS3ServerUrl = new Uri(s3Settings.CustomS3ServerUrl);
+            }
```
Comment (ayende) [inline src/Raven.Server/Documents/PeriodicBackup/Aws/RavenAwsS3Client.cs]: This seems to indicate that the region name isn't required for custom S3, but is that the case?
I don't think that this is true, see: https://developers.digitalocean.com/documentation/spaces/

File: src/Raven.Client/Documents/Operations/Backups/BackupSettings.cs [hunk 2]
```diff
@@ -125,10 +125,15 @@ public override DynamicJsonValue ToJson()
...
         /// <summary>
-        /// Amazon S3 Bucket name.
+        /// S3 Bucket name.
         /// </summary>
         public string BucketName { get; set; }
 
+        /// <summary>
+        /// S3 server Url when using custom server
+        /// </summary>
+        public string CustomS3ServerUrl { get; set; }
```
Comment (ppekrol) [inline src/Raven.Client/Documents/Operations/Backups/BackupSettings.cs]: We are in S3 Settings, so CustomServerUrl is enough.

File: src/Raven.Server/Documents/PeriodicBackup/Aws/RavenAwsS3Client.cs [hunk 3]
```diff
@@ -33,6 +33,10 @@ public class RavenAwsS3Client : RavenAwsClient
 
         private readonly string _bucketName;
         private readonly Logger _logger;
+        private readonly Uri _customS3ServerUrl;
+
+
+        private bool IsCustomS3Server => _customS3ServerUrl != null;
```
Comment (ppekrol) [inline src/Raven.Server/Documents/PeriodicBackup/Aws/RavenAwsS3Client.cs]: HasCustomServerUrl?

File: src/Raven.Server/Documents/PeriodicBackup/Aws/RavenAwsS3Client.cs [hunk 4]
```diff
@@ -927,12 +945,16 @@ public string CalculateMD5Hash(string input)
...
-            return $"{baseUrl}/{_bucketName}";
+            return IsCustomS3Server 
+                ? $"{_customS3ServerUrl}{_bucketName}" 
+                : $"{base.GetUrl()}/{_bucketName}";
         }
 
         public override string GetHost()
         {
+            if (IsCustomS3Server)
+                return _customS3ServerUrl.GetComponents(UriComponents.Host | UriComponents.Port, UriFormat.UriEscaped);
```
Comment (ppekrol) [inline src/Raven.Server/Documents/PeriodicBackup/Aws/RavenAwsS3Client.cs:953]: IIRC everything needs to be on HTTPS, but is this the case with custom server? Maybe someone will want a local instance without HTTPS? If we do want to handle that, then we need to return the protocol here also.

File: test/SlowTests/Server/Documents/PeriodicBackup/Restore/RestoreFromAwsS3.cs [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+        protected async Task incremental_and_full_check_last_file_for_backup() => await incremental_and_full_check_last_file_for_backup_internal();
+        [AmazonS3Fact, Trait("Category", "Smuggler")]
+        protected async Task incremental_and_full_backup_encrypted_db_and_restore_to_encrypted_DB_with_provided_key() => 
+            await incremental_and_full_backup_encrypted_db_and_restore_to_encrypted_DB_with_provided_key_internal();
+        [AmazonS3Fact, Trait("Category", "Smuggler")]
+        protected async Task snapshot_encrypted_db_and_restore_to_encrypted_DB() => await snapshot_encrypted_db_and_restore_to_encrypted_DB_internal();
+
+        protected override S3Settings GetS3Settings(string subPath = null)
+        {
+            var s3Settings = CustomS3FactAttribute.S3Settings;
```
Comment (ppekrol) [inline test/SlowTests/Server/Documents/PeriodicBackup/Restore/RestoreFromAwsS3.cs]: AmazonS3FactAttribute

File: test/Tests.Infrastructure/CustomS3FactAttribute.cs [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+            {
+                ParsingError = e.ToString();
+            }
+        }
+
+        public CustomS3FactAttribute([CallerMemberName] string memberName = "")
+        {
+            if (string.IsNullOrEmpty(ParsingError) == false)
+            {
+                Skip = $"Failed to parse S3 settings, error: {ParsingError}";
```
Comment (ppekrol) [inline test/Tests.Infrastructure/CustomS3FactAttribute.cs]: custom S3 settings

File: test/Tests.Infrastructure/CustomS3FactAttribute.cs [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+                return;
+            }
+
+            if (S3Settings == null)
+            {
+                Skip = $"S3 {memberName} tests missing S3 settings.";
```
Comment (ppekrol) [inline test/Tests.Infrastructure/CustomS3FactAttribute.cs:42]: custom S3 settings

File: src/Raven.Server/Documents/PeriodicBackup/Aws/RavenAwsS3Client.cs (hunk 4 above)
Comment (haludi) [reply src/Raven.Server/Documents/PeriodicBackup/Aws/RavenAwsS3Client.cs:953]: Actually, I check it so far just on HTTP. The `GetHost` function is used just for the header and it does not accept the schema.

**Pull request merged**